File: app/src/main/java/com/tablet/hid/util/GamepadLayoutResolver.kt
package com.tablet.hid.util

import android.content.Context
import org.json.JSONObject

data class GamepadElementRect(val left: Float, val top: Float, val w: Float, val h: Float)

object GamepadLayoutResolver {

    private var cachedJson: JSONObject? = null

    fun resolveLayout(context: Context, canvasW: Float, canvasH: Float): Map<String, GamepadElementRect> {
        val json = cachedJson ?: run {
            val text = context.assets.open("gamepad_layout.json").bufferedReader().use { it.readText() }
            JSONObject(text).also { cachedJson = it }
        }

        val topOffset = json.optDouble("topReservedDp", 0.0).toFloat()
        val elsJson = json.getJSONObject("elements")
        val chainGroupsJson = json.optJSONObject("chainGroups")

        val elIds = elsJson.keys().asSequence().toList()
        val ws = mutableMapOf<String, Float>()
        val hs = mutableMapOf<String, Float>()
        val xPos = mutableMapOf<String, Float>()
        val yPos = mutableMapOf<String, Float>()

        for (id in elIds) {
            val el = elsJson.getJSONObject(id)
            ws[id] = el.optDouble("w", 0.0).toFloat()
            hs[id] = el.optDouble("h", 0.0).toFloat()
        }

        // Pre-pass: resolve chain group x positions
        if (chainGroupsJson != null) {
            for (groupId in chainGroupsJson.keys()) {
                val group = chainGroupsJson.getJSONObject(groupId)
                val groupEls = group.getJSONArray("elements")
                val gap = group.optDouble("gap", 0.0).toFloat()
                var totalW = 0f
                for (i in 0 until groupEls.length()) totalW += ws[groupEls.getString(i)] ?: 0f
                totalW += gap * (groupEls.length() - 1)
                var x = if (group.optString("horizontalAlign") == "center") (canvasW - totalW) / 2f else 0f
                for (i in 0 until groupEls.length()) {
                    val id = groupEls.getString(i)
                    xPos[id] = x
                    x += (ws[id] ?: 0f) + gap
                }
            }
        }

        fun getXDeps(id: String): List<String> {
            if (xPos.containsKey(id)) return emptyList()
            val el = elsJson.getJSONObject(id)
            val deps = mutableListOf<String>()
            el.optString("startToStart").takeIf { it.isNotEmpty() && it != "parent" }?.let { deps += it }
            el.optString("startToEnd").takeIf { it.isNotEmpty() && it != "parent" }?.let { deps += it }
            el.optString("endToEnd").takeIf { it.isNotEmpty() && it != "parent" }?.let { deps += it }
            el.optString("endToStart").takeIf { it.isNotEmpty() && it != "parent" }?.let { deps += it }
            return deps
        }

        fun getYDeps(id: String): List<String> {
            val el = elsJson.getJSONObject(id)
            val deps = mutableListOf<String>()
            el.optString("topToTop").takeIf { it.isNotEmpty() && it != "parent" }?.let { deps += it }
            el.optString("topToBottom").takeIf { it.isNotEmpty() && it != "parent" }?.let { deps += it }
            el.optString("bottomToBottom").takeIf { it.isNotEmpty() && it != "parent" }?.let { deps += it }
            el.optString("bottomToTop").takeIf { it.isNotEmpty() && it != "parent" }?.let { deps += it }
            return deps
        }

        fun topoSort(ids: List<String>, getDeps: (String) -> List<String>): List<String> {
            val visited = mutableSetOf<String>()
            val order = mutableListOf<String>()
            fun visit(id: String) {
                if (id in visited) return
                visited += id
                for (dep in getDeps(id)) visit(dep)
                order += id
            }
            ids.forEach { visit(it) }
            return order
        }

        val xOrder = topoSort(elIds, ::getXDeps)
        val yOrder = topoSort(elIds, ::getYDeps)

        for (id in xOrder) {
            if (xPos.containsKey(id)) continue
            val el = elsJson.getJSONObject(id)
            val sm = el.optDouble("startMargin", 0.0).toFloat()
            val em = el.optDouble("endMargin", 0.0).toFloat()
            val w = ws[id] ?: 0f
            val sTS = el.optString("startToStart")
            val sTE = el.optString("startToEnd")
            val eTE = el.optString("endToEnd")
            val eTS = el.optString("endToStart")
            xPos[id] = when {
                sTS == "parent"      -> sm
                sTS.isNotEmpty()     -> (xPos[sTS] ?: 0f) + sm
                sTE == "parent"      -> canvasW + sm
                sTE.isNotEmpty()     -> (xPos[sTE] ?: 0f) + (ws[sTE] ?: 0f) + sm
                eTE == "parent"      -> canvasW - w - em
                eTE.isNotEmpty()     -> (xPos[eTE] ?: 0f) + (ws[eTE] ?: 0f) - w - em
                eTS == "parent"      -> -w - em
                eTS.isNotEmpty()     -> (xPos[eTS] ?: 0f) - w - em
                else                 -> 0f
            }
        }

        for (id in yOrder) {
            val el = elsJson.getJSONObject(id)
            val tm = el.optDouble("topMargin", 0.0).toFloat()
            val bm = el.optDouble("bottomMargin", 0.0).toFloat()
            val h = hs[id] ?: 0f
            val tTT = el.optString("topToTop")
            val tTB = el.optString("topToBottom")
            val bTB = el.optString("bottomToBottom")
            val bTT = el.optString("bottomToTop")
            yPos[id] = when {
                tTT == "parent"      -> topOffset + tm
                tTT.isNotEmpty()     -> (yPos[tTT] ?: 0f) + tm
                tTB == "parent"      -> canvasH + tm
                tTB.isNotEmpty()     -> (yPos[tTB] ?: 0f) + (hs[tTB] ?: 0f) + tm
                bTB == "parent"      -> canvasH - h - bm
                bTB.isNotEmpty()     -> (yPos[bTB] ?: 0f) + (hs[bTB] ?: 0f) - h - bm
                bTT == "parent"      -> -h - bm
                bTT.isNotEmpty()     -> (yPos[bTT] ?: 0f) - h - bm
                else                 -> 0f
            }
        }

        return elIds.associateWith { id ->
            GamepadElementRect(
                left = xPos[id] ?: 0f,
                top = yPos[id] ?: 0f,
                w = ws[id] ?: 0f,
                h = hs[id] ?: 0f,
            )
        }
    }
}

File: ios/TabletHID/Services/GamepadLayoutResolver.swift
import Foundation

struct GamepadElementRect {
    let left: Double
    let top: Double
    let w: Double
    let h: Double
}

/// Resolves the natural (un-offset) position of every gamepad element on a given canvas.
///
/// Mirrors GamepadLayoutResolver.kt exactly — reads gamepad_layout.json from the app bundle,
/// then performs the same dependency-sorted constraint pass to produce absolute rects.
enum GamepadLayoutResolver {

    private static var cachedJson: [String: Any]?

    /// Returns a map from element ID to its natural rect on a canvas of size (canvasW × canvasH).
    static func resolveLayout(canvasW: Double, canvasH: Double) -> [String: GamepadElementRect] {
        let json: [String: Any]
        if let cached = cachedJson {
            json = cached
        } else if
            let url  = Bundle.main.url(forResource: "gamepad_layout", withExtension: "json"),
            let data = try? Data(contentsOf: url),
            let parsed = try? JSONSerialization.jsonObject(with: data) as? [String: Any]
        {
            cachedJson = parsed
            json = parsed
        } else {
            return [:]
        }

        let topOffset = json["topReservedDp"] as? Double ?? 0.0
        guard let elsJson = json["elements"] as? [String: [String: Any]] else { return [:] }
        let chainGroupsJson = json["chainGroups"] as? [String: [String: Any]]

        let elIds = Array(elsJson.keys)
        var ws   = [String: Double]()
        var hs   = [String: Double]()
        var xPos = [String: Double]()
        var yPos = [String: Double]()

        for id in elIds {
            let el = elsJson[id]!
            ws[id] = el["w"] as? Double ?? 0.0
            hs[id] = el["h"] as? Double ?? 0.0
        }

        // Pre-pass: resolve chain-group horizontal positions.
        if let chainGroupsJson = chainGroupsJson {
            for (_, group) in chainGroupsJson {
                guard let groupEls = group["elements"] as? [String] else { continue }
                let gap = group["gap"] as? Double ?? 0.0
                var totalW = 0.0
                for id in groupEls { totalW += ws[id] ?? 0.0 }
                totalW += gap * Double(groupEls.count - 1)
                var x = (group["horizontalAlign"] as? String == "center") ? (canvasW - totalW) / 2.0 : 0.0
                for id in groupEls {
                    xPos[id] = x
                    x += (ws[id] ?? 0.0) + gap
                }
            }
        }

        func xDeps(of id: String) -> [String] {
            if xPos[id] != nil { return [] }
            guard let el = elsJson[id] else { return [] }
            return ["startToStart", "startToEnd", "endToEnd", "endToStart"].compactMap { key in
                let v = el[key] as? String ?? ""
                return (!v.isEmpty && v != "parent") ? v : nil
            }
        }

        func yDeps(of id: String) -> [String] {
            guard let el = elsJson[id] else { return [] }
            return ["topToTop", "topToBottom", "bottomToBottom", "bottomToTop"].compactMap { key in
                let v = el[key] as? String ?? ""
                return (!v.isEmpty && v != "parent") ? v : nil
            }
        }

        func topoSort(_ ids: [String], deps: (String) -> [String]) -> [String] {
            var visited = Set<String>()
            var order   = [String]()
            func visit(_ id: String) {
                guard !visited.contains(id) else { return }
                visited.insert(id)
                for dep in deps(id) { visit(dep) }
                order.append(id)
            }
            ids.forEach { visit($0) }
            return order
        }

        for id in topoSort(elIds, deps: xDeps) {
            if xPos[id] != nil { continue }
            guard let el = elsJson[id] else { continue }
            let sm = el["startMargin"] as? Double ?? 0.0
            let em = el["endMargin"]   as? Double ?? 0.0
            let w  = ws[id] ?? 0.0
            let sTS = el["startToStart"] as? String ?? ""
            let sTE = el["startToEnd"]   as? String ?? ""
            let eTE = el["endToEnd"]     as? String ?? ""
            let eTS = el["endToStart"]   as? String ?? ""
            if      sTS == "parent" { xPos[id] = sm }
            else if !sTS.isEmpty    { xPos[id] = (xPos[sTS] ?? 0.0) + sm }
            else if sTE == "parent" { xPos[id] = canvasW + sm }
            else if !sTE.isEmpty    { xPos[id] = (xPos[sTE] ?? 0.0) + (ws[sTE] ?? 0.0) + sm }
            else if eTE == "parent" { xPos[id] = canvasW - w - em }
            else if !eTE.isEmpty    { xPos[id] = (xPos[eTE] ?? 0.0) + (ws[eTE] ?? 0.0) - w - em }
            else if eTS == "parent" { xPos[id] = -w - em }
            else if !eTS.isEmpty    { xPos[id] = (xPos[eTS] ?? 0.0) - w - em }
            else                    { xPos[id] = 0.0 }
        }

        for id in topoSort(elIds, deps: yDeps) {
            guard let el = elsJson[id] else { continue }
            let tm = el["topMargin"]    as? Double ?? 0.0
            let bm = el["bottomMargin"] as? Double ?? 0.0
            let h  = hs[id] ?? 0.0
            let tTT = el["topToTop"]       as? String ?? ""
            let tTB = el["topToBottom"]    as? String ?? ""
            let bTB = el["bottomToBottom"] as? String ?? ""
            let bTT = el["bottomToTop"]    as? String ?? ""
            if      tTT == "parent" { yPos[id] = topOffset + tm }
            else if !tTT.isEmpty    { yPos[id] = (yPos[tTT] ?? 0.0) + tm }
            else if tTB == "parent" { yPos[id] = canvasH + tm }
            else if !tTB.isEmpty    { yPos[id] = (yPos[tTB] ?? 0.0) + (hs[tTB] ?? 0.0) + tm }
            else if bTB == "parent" { yPos[id] = canvasH - h - bm }
            else if !bTB.isEmpty    { yPos[id] = (yPos[bTB] ?? 0.0) + (hs[bTB] ?? 0.0) - h - bm }
            else if bTT == "parent" { yPos[id] = -h - bm }
            else if !bTT.isEmpty    { yPos[id] = (yPos[bTT] ?? 0.0) - h - bm }
            else                    { yPos[id] = 0.0 }
        }

        return Dictionary(uniqueKeysWithValues: elIds.map { id in
            (id, GamepadElementRect(
                left: xPos[id] ?? 0.0,
                top:  yPos[id] ?? 0.0,
                w:    ws[id]   ?? 0.0,
                h:    hs[id]   ?? 0.0
            ))
        })
    }
}
